pkg/providers/cloudstack/cloudstack.go

```diff
@@ -853,6 +1075,84 @@ func (p *cloudstackProvider) GenerateStorageClass() []byte {
 	return nil
 }
 
+func (p *cloudstackProvider) setupSSHAuthKeysForUpgrade() error {
+	var err error
+	controlPlaneUser := p.machineConfigs[p.clusterConfig.Spec.ControlPlaneConfiguration.MachineGroupRef.Name].Spec.Users[0]
+	p.controlPlaneSshAuthKey = controlPlaneUser.SshAuthorizedKeys[0]
+	if len(p.controlPlaneSshAuthKey) > 0 {
+		p.controlPlaneSshAuthKey, err = common.StripSshAuthorizedKeyComment(p.controlPlaneSshAuthKey)
+		if err != nil {
+			return err
+		}
+	}
+	controlPlaneUser.SshAuthorizedKeys[0] = p.controlPlaneSshAuthKey
+	for _, workerNodeGroupConfiguration := range p.clusterConfig.Spec.WorkerNodeGroupConfigurations {
+		workerUser := p.machineConfigs[workerNodeGroupConfiguration.MachineGroupRef.Name].Spec.Users[0]
+		p.workerSshAuthKey = workerUser.SshAuthorizedKeys[0]
+		if len(p.workerSshAuthKey) > 0 {
+			p.workerSshAuthKey, err = common.StripSshAuthorizedKeyComment(p.workerSshAuthKey)
+			if err != nil {
+				return err
+			}
+		}
+		workerUser.SshAuthorizedKeys[0] = p.workerSshAuthKey
+	}
+	if p.clusterConfig.Spec.ExternalEtcdConfiguration != nil {
+		etcdUser := p.machineConfigs[p.clusterConfig.Spec.ExternalEtcdConfiguration.MachineGroupRef.Name].Spec.Users[0]
+		p.etcdSshAuthKey = etcdUser.SshAuthorizedKeys[0]
+		if len(p.etcdSshAuthKey) > 0 {
+			p.etcdSshAuthKey, err = common.StripSshAuthorizedKeyComment(p.etcdSshAuthKey)
+			if err != nil {
+				return err
+			}
+		}
+		etcdUser.SshAuthorizedKeys[0] = p.etcdSshAuthKey
+	}
```

Comment:
could we do all of this by iterating over `p.clusterConfig.MachineConfigRefs()`?

Reply:
I'd be a bit nervous to do this. It seems like we're at least setting the internal attributes `p.controlPlaneSshAuthKey`, `p.etcdSshAuthKey`, and `p.workerSshAuthKey` explicitly which we won't be able to do if we're just iterating, which are being used elsewhere. Otherwise we'd have to track which machine config we're looking at on each iteration which seems to defeat the purpose

I guess we could iterate over the machineconfigrefs and build a map between machineconfigref name -> auth key but then it's still not clear how to go back from that map to assign back to p.etcdSshAuthKey, for example. I'd be inclined to wait on this work

Reply:
Where are the ssh key fields used in the provider? For Tinkerbell the idea of populating ssh key state has been stripped out because it made no sense. We actually iterate over machine configs (albeit using some other state) as suggested by @g-gaston because the ssh key generation logic was remarkably convoluted.

Specifically, we found object state being used to communicate between method calls on a single call path - i.e. they should have been parameters. Tinkerbell still isn't right but it simplifies the ssh key generation problem. We hope to make that logic re-usable soon.

Reply:
If it's going to be made reusable soon, would it make sense to leave the current logic as is (since it works and has been tested) and wait for that functionality to be made reusable? Especially since it's the same across providers

The ssh key fields are used in capi spec generation

Reply:
What is the need for `p.controlPlaneSshAuthKey`, `p.etcdSshAuthKey`, and `p.workerSshAuthKey`?
They all seem to be extracted from `p.machineConfigs`, which is also a provider member, hence available everywhere. So first sight, they seem redundant. I could be missing something here.

Reply:
Removing the attributes in favor of using p.machineConfigs directly